### PR TITLE
Encode PromQL Matricies with Arrow

### PR DESCRIPTION
**Pull request by csmarchbanks**
Use content negotiation to return a stream of [Arrow](https://arrow.apache.org/) records from the query and query_range endpoints in the case where a matrix is returned. This provides vastly superior CPU performance both on the server and on the client when transferring significant amounts of data.

This is a very minimal implementation, and only targets the matrix type of response to begin with as that is the only response that can return very large amounts of data. I try to explain the response, and leave it open to implementing streaming samples for each series, but today all samples are returned in a single record batch for performance reasons (and the 11,000 point limit makes this safe).

Happy to iterate on the format or code. I also wonder if we should put this behind a feature flag, but since content negotiation is required an additional flag did not feel necessary to me. The documentation still calls out the arrow type as experimental. More information on the Arrow Streaming format can be found at https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format.

```
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/web/api/v1
BenchmarkRespond-8                  2326            475496 ns/op          213687 B/op          8 allocs/op
BenchmarkRespondArrow-8            16176             74459 ns/op          452520 B/op         83 allocs/op
```

### CODE REVIEW

**Comment by roidelapluie** [comment]
Could the conversion to arrow happen in a dedicated package, instead of web?

File: web/api/v1/api_test.go [hunk 1]
```diff
@@ -3300,10 +3369,35 @@ func BenchmarkRespond(b *testing.B) {
...
+		ResultType: parser.ValueTypeMatrix,
+		Result: promql.Matrix{
+			promql.Series{
+				Points: points,
+				Metric: nil,
+			},
+		},
+	}
+	request, _ := http.NewRequest("GET", "http://localhost:9090", nil)
+	request.Header.Add("Accept", "application/vnd.apache.arrow.stream")
```
Comment (bwplotka) [inline web/api/v1/api_test.go:3396]: Can we add constants for the Accept header value? (and name)

File: web/api/v1/api.go [hunk 2]
```diff
@@ -1526,7 +1530,70 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
...
 }
 
-func (api *API) respond(w http.ResponseWriter, data interface{}, warnings storage.Warnings) {
+func (api *API) tryArrow(w http.ResponseWriter, r *http.Request, data interface{}, warnings storage.Warnings) error {
+	result, ok := data.(*queryData)
+	if !ok {
+		return fmt.Errorf("invalid result data type: %T", data)
+	}
+
+	if result.ResultType != parser.ValueTypeMatrix {
```
Comment (bwplotka) [inline web/api/v1/api.go:1539]: As @roidelapluie mentioned, this could be done in some helper function in promQL package (?)?

**Review by bwplotka — DISMISSED** [review]
Nice! Good start IMO 💪🏽 

Tiny comments only.

File: web/api/v1/api_test.go (hunk 1 above)
Comment (csmarchbanks) [reply web/api/v1/api_test.go:3396]: I moved the MIME type to a const, but left the header name as a const string as that seems to be what we do elsewhere in this package and Accept and Content-Type are never going to change. Does that seem reasonable?

**Comment by csmarchbanks** [comment]
I added a commit that moves the arrow transformation and writes to the Matrix and Series types in the promql package and hid them behind an interface so we could support Vector or other types in the future with minimal changes. What do you think of that approach? I am also happy to move the code to a separate arrow package if desired, perhaps as a sub-package of promql?

File: promql/value.go [hunk 3]
```diff
@@ -101,6 +107,34 @@ func (s Series) MarshalJSON() ([]byte, error) {
 	return json.Marshal(series)
 }
 
+func (s Series) writeArrow(w io.Writer, pool memory.Allocator) error {
+	fields := []arrow.Field{
+		{Name: "t", Type: &arrow.TimestampType{Unit: arrow.Millisecond}},
+		{Name: "v", Type: arrow.PrimitiveTypes.Float64},
+	}
+	metadata := arrow.MetadataFrom(s.Metric.Map())
```
Comment (metalmatze) [inline promql/value.go:115]: Is there a specific reason to put the metric as map into the metadata? 
I probably would have put it into the schema itself. I can see both work, to be honest.

**Review by metalmatze — COMMENTED** [review]
Very cool! :rocket:

File: promql/value.go (hunk 3 above)
Comment (csmarchbanks) [reply promql/value.go:115]: The reason for using Metadata is mostly how we use it internally now for transferring time series data between a few services. I guess the other option would be to add it as the name of the value column and have the client parse it as JSON?

**Comment by alanprot** [comment]
This is amazing..

I wonder if we should propose a new streamed QueryAPI (We should be able to stream by series like here or by steps?).

Prometheus compatibles services like Thanos and Cortex already split one single query in multiple ones (by day for instance) but we need to wait all subqueries to finish to return the result back to the customer. I think i better UX would be to load the graphs as soon as parts of the results are available.

**Comment by roidelapluie** [comment]
> This is amazing..
> 
> I wonder if we should propose a new streamed QueryAPI (We should be able to stream by series like here or by steps?).
> 
> Prometheus compatibles services like Thanos and Cortex already split one single query in multiple ones (by day for instance) but we need to wait all subqueries to finish to return the result back to the customer. I think i better UX would be to load the graphs as soon as parts of the results are available.

I disagree as working with partial data, or thinking your graph is complete when it is not is really confusing and difficult to work with.

**Comment by alanprot** [comment]
> > This is amazing..
> > I wonder if we should propose a new streamed QueryAPI (We should be able to stream by series like here or by steps?).
> > Prometheus compatibles services like Thanos and Cortex already split one single query in multiple ones (by day for instance) but we need to wait all subqueries to finish to return the result back to the customer. I think i better UX would be to load the graphs as soon as parts of the results are available.
> 
> I disagree as working with partial data, or thinking your graph is complete when it is not is really confusing and difficult to work with.

This is not a UX problem? we can hint that the data is "loading"... and it would be up the the consumer to ask partial data or not. The problem we see is that sometimes the response can be quite big and consume lots of resource just to serialize/deserialize.

**Comment by roidelapluie** [comment]
> This is not a UX problem? we can hint that the data is "loading"... and it would be up the the consumer to ask partial data or not. The problem we see is that sometimes the response can be quite big and consume lots of resource just to serialize/deserialize.

Grafana solved this by only requesting the last steps when it already knows part of the graph (e.g. when refreshing).

**Comment by csmarchbanks** [comment]
At the dev-summit after PromCon we did agree that a streaming API is something we are interested in. However, I do not want to derail this pull request with discussing a streaming API as supporting streaming responses in the PromQL engine is a far larger change than this.

However, I do want the wire format in this PR to support streaming use cases, and the question of if we want to stream by series and samples (what is present in this PR), or stream by timestamp with all series in one row is relevant to the wire format of this PR. For example, rather than the simple schema I use here (two columns, `t` and `v`) I could add every series as a separate column with one `t` column in addition. I am experimenting a bit with that format, but I am not sure I love it as you can end up with a fairly sparse matrix in the presence of churn or an instant query like `up[30d]`. I would love to hear the opinions of others on per-series vs per-timestamp streaming though.

Plus, if the wire format supports streaming, those other projects can experiment with/implement streaming as they want :).

File: web/api/v1/api.go [hunk 4]
```diff
@@ -1526,7 +1529,42 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
...
+	w.WriteHeader(http.StatusOK)
+	if err := arrowWriter.WriteArrow(w); err != nil {
+		level.Error(api.logger).Log("msg", "error writing arrow response", "err", err)
+	}
+	return nil
+}
+
+func (api *API) respond(w http.ResponseWriter, r *http.Request, data interface{}, warnings storage.Warnings) {
+	if r.Header.Get("Accept") == mimeTypeArrowStream {
+		err := api.tryArrowResponse(w, r, data, warnings)
```
Comment (LeviHarrison) [inline web/api/v1/api.go:1562]: Why return full error messages and errors at all if nothing is going to be done with them?

**Comment by roidelapluie** [comment]
I was more thinking about moving the modification to the `model/` package than into promql itself, would it make sense for you?

File: web/api/v1/api.go (hunk 4 above)
Comment (csmarchbanks) [reply web/api/v1/api.go:1562]: Oops, I meant to log them at a debug level. Added the log now.

**Comment by csmarchbanks** [comment]
> I was more thinking about moving the modification to the model/ package than into promql itself, would it make sense for you? It would then be usable for third parties promql engines as well.

That seems reasonable, I will at least investigate it next week!

File: web/api/v1/api.go (hunk 4 above)
Comment (LeviHarrison) [reply web/api/v1/api.go:1562]: Cool thanks

**Comment by brancz** [comment]
Is there a particular reason we’re doing this in the query range API? I personally find this an odd placement and usage of arrow, typically arrow the format used within a query engine for zero copy computations. Have we considered doing this rather for remote read? That seems like the more natural space for it to me. And streaming there would also work very well as we already have machinery available from the proto streaming variant.

**Comment by csmarchbanks** [comment]
Yes, the idea of supporting Arrow in the query(_range) API is to better support data analytics and machine learning use cases. For those use cases I have found it common to at least use a function like `rate` and commonly also some level of aggregation. I would not want a client to have to deal with things like counter resets that are handled by PromQL which means remote read does not work well for these use cases.

**Comment by brancz** [comment]
Got it. That makes sense. I hope that eventually, we have PromQL be arrow-based, and then it doesn't really matter, but until then I agree it's difficult to make a case for just remote-read. Thanks for elaborating!

**Comment by krajorama** [comment]
> For my use case it would be nice if the format of the (float) histogram results returned by such API is the encoded [internal representation](https://github.com/prometheus/prometheus/tree/main/model/histogram) and not what Prometheus returns currently which is the [user-friendly JSON](https://github.com/prometheus/prometheus/blob/72f20d949afa2e960b5003a952c8390493252716/web/api/v1/api.go#L1842).

Actually to avoid favoring one use-case, it would be nice to be able to dependency inject additional serializers into the web/api. One could select from client side with [Accept](https://www.rfc-editor.org/rfc/rfc7231#section-5.3.2) header which one to use. This would enable me to add my own serializer and get the data in the format that suits me.

**Comment by csmarchbanks** [comment]
> For my use case it would be nice if the format of the (float) histogram results returned by such API is the encoded [internal representation](https://github.com/prometheus/prometheus/tree/main/model/histogram) and not what Prometheus returns currently which is the [user-friendly JSON](https://github.com/prometheus/prometheus/blob/72f20d949afa2e960b5003a952c8390493252716/web/api/v1/api.go#L1842).

I agree, as you can see I have not implemented native histograms yet, but my thought is to have a set of columns that represent the internal structure for native histograms. I looked briefly at using the JSON representation but performance was significantly worse than I would have liked.

I am planning to come back to this PR soon. Talking with Julien we decided it would be best to move the translation back to web for now.

**Comment by beorn7** [comment]
WRT native histograms: While the internal representation is very efficient, it requires very specific knowledge to decode.

If the goal of this Arrow approach is to shovel vast amounts of metrics data into some 3rd party data processing system (as @csmarchbanks said above: "support data analytics and machine learning use cases"), it might be better to still use a more explicit representation of the native histograms (more or less something like the JSON representation, just not in JSON). Otherwise, you need to reproduce the decoding logic that is currently "internal to Prometheus" in the 3rd party data processing system (and worse, you also have to keep it up to date with "Prometheus internal" changes, e.g. adding new bucket layout schemas).

If the goal is to transfer the data into another system that is still based on how Prometheus internally works (that's @krajorama's use case), then using the internal "encoded" version of the native histograms is fine.

File: web/api/v1/api.go [hunk 5]
```diff
@@ -1548,7 +1554,71 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
...
+	w.WriteHeader(http.StatusOK)
+
+	pool := memory.NewGoAllocator()
+	fields := []arrow.Field{
+		{Name: "t", Type: &arrow.TimestampType{Unit: arrow.Millisecond}},
+		{Name: "v", Type: arrow.PrimitiveTypes.Float64},
+	}
+
+	for _, series := range matrix {
+		metadata := arrow.MetadataFrom(series.Metric.Map())
```
Comment (brancz) [inline web/api/v1/api.go:1587]: Having worked with arrow quite a bit now, I think this is rather odd. I would have expected labels to be a column of type map, and then each row has a list of timestamps and values. That way we can also pack multiple series into a single record.

**Comment by roidelapluie** [comment]
@csmarchbanks I think it's still a really great thing to have. Would you please rebase and take @brancz 's suggestion into consideration? Thanks :-)

**Comment by brancz** [comment]
I’ve thought about this even more and I think I’ve come to the conclusion that samples should be a list of a struct that under the hood is just a column for timestamp and value. Happy to sketch out the schema if you want!

**Comment by beorn7** [comment]
Looking at this during the bug scrub.

@csmarchbanks is this still on your radar? The idea seems to be universally loved.

**Comment by csmarchbanks** [comment]
Yep, this is still on my radar, but I won't have time to work on it until at least after my second batch of paternity leave (end of January).

I would be happy to see some additional schema ideas to discuss, I am a little bit hesitant to go too far into custom struct types as I want this to be easy to use with things like pandas in the data science realm, but I could also provide a library to make that easier.

**Comment by brancz** [comment]
Nothing I suggested is custom, I think serializing the labels as metadata as it's proposed in this PR is very custom.

I think a dictionary + RLE encoded string column for each label-name and a list of structs that's timestamp and value would be the most idiomatic arrow schema.

**Comment by csmarchbanks** [comment]
I'll rephrase that, my motivation for the format in this PR was to make it as easy and efficient as possible to operate on those time series using tools like Pandas or NumPy. In my experience all of those tools use a column for timestamp and then one or more columns for values to represent a time series. You should be able to directly load this Arrow schema into a Pandas series without any transformations or allocations. By "custom" I meant types that I think might make it harder to load directly into Pandas/Numpy (see https://arrow.apache.org/docs/python/pandas.html#zero-copy-series-conversions). 

Anyway, I do like that your proposed schema can add multiple series to the same record batch and avoids using metadata fields. I also like that it would work well for a `vector` response whereas my original proposal would be a massive amount of overhead. However, I would like to discuss some of the motivation/tradeoffs for each. I created a quick [design doc](https://arrow.apache.org/docs/python/pandas.html#zero-copy-series-conversions) where we can do this, would you be willing to add to/correct the details I added for your proposal?

Basically, I am happy to implement the format you proposed when I have time to come back to this, but would like to discuss the motivation and tradeoffs first.

**Comment by brancz** [comment]
Thank you so much for putting that together! I commented with a few thoughts. I think there are a few things to explore, but loving where this is going. (btw at this point we should discuss having an arrow flight compatible API)

**Comment by bboreham** [comment]
Hello again from the bug-scrub!  @csmarchbanks could you give us another status update please?

We noted that content-type negotiation has become more powerful, which might make this PR easier (and requires a rebase).

**Comment by csmarchbanks** [comment]
I am going to close this PR as I don't have time to finish it off.

The main point of friction is that it is undecided how the arrow result should be returned. There isn't a great answer as different users seem to want different representations. For my ML workflows I basically wanted column based and streaming to make it trivial to import Prometheus query results into [numpy](https://pandas.pydata.org/docs/user_guide/timeseries.html) for futher analysis. However, for many other machine to machine use cases Prometheus typically returns data in a more row (series) oriented manner which is what @brancz expected the output to look like.

A lot of the performance issues could also be solved by using a protobuf encoding, but with the same result structure used by the existing JSON API. Protobuf would have fewer decisions about how to represent the data and still allow more efficient machine to machine communication. In fact, we could likely just take what was done in Mimir (see [mimir.proto](https://github.com/grafana/mimir/blob/5bdf5cd45eff0214eba898808c2beeb43c305a8d/pkg/mimirpb/mimir.proto#L253) and the `application/vnd.mimir.queryresponse+protobuf` content type) and make it an official content type.
